src/auth/externalclient.ts

```diff
@@ -0,0 +1,313 @@
+  GaxiosPromise,
+  GaxiosResponse,
+} from 'gaxios';
+import * as stream from 'stream';
+
+import {Credentials} from './credentials';
+import {AuthClient} from './authclient';
+import {BodyResponseCallback} from '../transporters';
+import {GetAccessTokenResponse, Headers} from './oauth2client';
+import * as sts from './stscredentials';
+import {ClientAuthentication} from './oauth2common';
+
+/**
+ * The required token exchange grant_type: rfc8693#section-2.1
+ */
+const STS_GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:token-exchange';
+/**
+ * The requested token exchange requested_token_type: rfc8693#section-2.1
+ */
+const STS_REQUEST_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:access_token';
+/** The default OAuth scope to request when none is provided. */
+const DEFAULT_OAUTH_SCOPE = 'https://www.googleapis.com/auth/cloud-platform';
+/**
+ * Offset to take into account network delays and server clock skews.
+ */
+export const EXPIRATION_TIME_OFFSET = 5 * 60 * 1000;
+/** The credentials JSON file type for external account clients. */
+const EXTERNAL_ACCOUNT_TYPE = 'external_account';
+
+/**
+ * Base external account credentials json interface.
+ */
+export interface ExternalAccountClientOptions {
+  type: string;
+  audience: string;
+  subject_token_type: string;
+  service_account_impersonation_url?: string;
+  token_url: string;
+  token_info_url?: string;
+  client_id?: string;
+  client_secret?: string;
+  credential_source: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [key: string]: any;
+  };
+  quota_project_id?: string;
+}
+
+/**
+ * Internal interface for tracking the access token expiration time.
+ */
+interface CredentialsWithResponse extends Credentials {
+  res?: GaxiosResponse | null;
+}
+
+/**
+ * Base external account client. This is used to instantiate AuthClients for
+ * exchanging external account credentials for GCP access token and authorizing
+ * requests to GCP APIs.
+ * The base class implements common logic for exchanging various type of
+ * external credentials for GCP access token. The logic of determining and
+ * retrieving the external credential based on the environment and
+ * credential_source will be left for the subclasses.
+ */
+export abstract class ExternalAccountClient extends AuthClient {
+  /**
+   * OAuth scopes for the GCP access token to use. When not provided,
+   * the default https://www.googleapis.com/auth/cloud-platform is
+   * used.
+   */
+  public scopes?: string | string[];
+  private cachedAccessToken: CredentialsWithResponse | null;
+  private readonly stsCredential: sts.StsCredentials;
+
+  /**
+   * Instantiate an ExternalAccountClient instance using the provided JSON
+   * object loaded from an external account credentials file.
+   * @param options The external account options object typically loaded
+   *   from the external account JSON credential file.
+   */
+  constructor(private readonly options: ExternalAccountClientOptions) {
```

Comment:
do you know if the `readonly` modifier applies to accessing properties on the object? I fixed a variety of bugs in this library related to people modifying an object, not realizing it's passed by reference.